# [E-02] Classfifcation_wine  

# 미니 프로젝트 (2) load_wine : 와인을 분류해 보자

## 첫 시도

cloud 환경이 아닌 local(m1 맥북) 환경에서 테이스 한 결과임을 참고 바란다.

sklearn wine dataset을 활용하여 분류해 보자.


### 1. 모듈 import

In [1]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

### 2. 데이터 준비  

- load_wine 메서드를 사용합니다.
- wines_data에 Feature data 지정
- wines_label에 Label(target) data 지정


In [2]:
wines = load_wine()
wines_data = wines.data
print(wines_data.shape, '\n')

wines_label = wines.target
print(wines_label.shape, '\n')
# print(wines_label[:20],'\n')
wines.target

(178, 13) 

(178,) 



array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

- wines_data에 178개의 데이터가 각각 13개의 정보를 담고 있다.
- wines_label에도 178개의 데이터가 있음을 확인
- wines.target에서 `0, 1, 2` 3개의 class가 있음을 확인



### 3. 데이터 이해하기

다루어야 할 데이터를 자세히 살펴보자.

- Target Names 출력해 보기
- 데이터 Describe 해 보기


In [3]:
print(wines.target_names)

['class_0' 'class_1' 'class_2']


- wines 데이터 셋의 Target Names을 확인
- 3개의 class로 분류됨을 확인

In [4]:
print(wines.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

- wines 데이터셋의 설명을 통해서도 기본적인 정보확인 가능
- `3개의 class : 0(59개), 1(71개), 2(48개) 확인`

### 4. 데이터 분리

In [5]:
X_train, X_test, y_train, y_test = train_test_split(wines_data,
                                                   wines_label,
                                                   test_size=0.2,
                                                   random_state=15)

print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test), '\n')
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

X_train 개수:  142 , X_test 개수:  36 

(142, 13) (142,)
(36, 13) (36,)


- wines 데이터 셋을 학습용 데이터와 테스트용 데이터로 분리
- test 데이터 셋의 크기는 전체 데이터셋의 20% 적용하여
- 학습용 데이터 142개, 테스트용 데이터 36개로 분리
- 데이터 분리할 때 랜덤으로 섞기 위해 random_state=15 적용

### 5. 다양한 모델로 학습시키기

1. Decision Tree 사용해 보기
2. Random Forest 사용해 보기
3. SVM 사용해 보기
4. SGD Classifier 사용해 보기
5. Logistic Regression 사용해 보기

In [6]:
# 각 모델 import
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

# 모델별 변수에 저장
decision_tree = DecisionTreeClassifier(random_state=15)
random_forest = RandomForestClassifier(random_state=15)
svm_model = svm.SVC()
sgd_model = SGDClassifier()
lr_model = LogisticRegression(max_iter=4000)  # digits작업시 warning message 해결방법 적용

### 5.1 Decision Tree

In [7]:
decision_tree.fit(X_train, y_train)
y_pred_decision_tree = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred_decision_tree))

              precision    recall  f1-score   support

           0       0.92      1.00      0.96        12
           1       0.91      0.83      0.87        12
           2       0.92      0.92      0.92        12

    accuracy                           0.92        36
   macro avg       0.92      0.92      0.92        36
weighted avg       0.92      0.92      0.92        36



### Decision Tree를 이용한 결과

- accuracy : 92%


### 5.2 Random Forest

In [8]:
random_forest.fit(X_train, y_train)
y_pred_random_forest = random_forest.predict(X_test)

print(classification_report(y_test, y_pred_random_forest))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



### Random Forest를 이용한 결과

- accuracy : 100%


### 5.3 SVM

In [9]:
svm_model.fit(X_train, y_train)
y_pred_svm_model = svm_model.predict(X_test)

print(classification_report(y_test, y_pred_svm_model))

              precision    recall  f1-score   support

           0       0.79      0.92      0.85        12
           1       0.50      0.92      0.65        12
           2       0.00      0.00      0.00        12

    accuracy                           0.61        36
   macro avg       0.43      0.61      0.50        36
weighted avg       0.43      0.61      0.50        36



/opt/homebrew/Caskroom/miniforge/base/envs/aiffel/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/homebrew/Caskroom/miniforge/base/envs/aiffel/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/homebrew/Caskroom/miniforge/base/envs/aiffel/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _wa

# SVM을 이용한 결과

- accuracy : 61%


### 5.4 SGD Classifier

In [10]:
sgd_model.fit(X_train, y_train)
y_pred_sgd_model = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred_sgd_model))

              precision    recall  f1-score   support

           0       0.65      0.92      0.76        12
           1       1.00      0.25      0.40        12
           2       0.50      0.67      0.57        12

    accuracy                           0.61        36
   macro avg       0.72      0.61      0.58        36
weighted avg       0.72      0.61      0.58        36



### SGD Classifier를 이용한 결과

- accuracy : 50%


### 5.5 Logistic Regression

In [11]:
lr_model.fit(X_train, y_train)
y_pred_lr_model = lr_model.predict(X_test)

print(classification_report(y_test, y_pred_lr_model))

              precision    recall  f1-score   support

           0       1.00      0.92      0.96        12
           1       0.86      1.00      0.92        12
           2       1.00      0.92      0.96        12

    accuracy                           0.94        36
   macro avg       0.95      0.94      0.95        36
weighted avg       0.95      0.94      0.95        36



### Logistic Regression을 이용한 결과

- accuracy : 94%

### 결론

- #### accuracy

    - Decision Tree : 92%
    - Random Forest : 100%
    - SVM : 61%
    - SGD Classifier : 50%
    - Logistic Regression : 94% 

- 모델의 성능을 평가하는 척도는 와인의 종류가 무엇인지 정확히 예측하는 것이기에 **정확도(accuracy)**를 이용하는 것이 적합한 것으로 생각된다.
- SVM과 SGD Classifier에서는 나머지 모델들에 비해 현저히 낮은 accuracy를 보여줌
    `(warning message와 연관성이 있어보임)`

    - 이것은 class 2에 대한 f1-score가 표현되지 못하면서 낮아진 것으로 판단된다.
    - 이유는 class 2에 대해 f1-score가 없기때문에 0.0으로 간주되어서 그렇다고 한다.
      구글링을 통한 결과는 이와 같았다.
    - 왜 이런지는 project 제출 시간에 임박하여 추후 알아보도록 하려고 한다.


- SVM과 SGD Classifier를 제외한 나머지 3개의 모델을 이용하는 것이 와인 분류에 적합하다고 판단
- 특히, Random Forest의 경우 **100%**라는 믿기 어려운 accuracy를 보여줌
- 따라서 해당 프로젝트에서는 **Random Forest**을 사용하는 것이 가장 적합하다고 판단

**추가생각**  
**`Random Forest는 선형 모델에 비해 많은 메모리를 사용하며 훈련과 예측이 느리다.`**  
**`속도와 메모리 사용에 제약이 있는 어플리케이션이라면 선형모델이 적합할 수 있겠다.`**